### ISSUE TITLE
Named Lights' Lightmap Pages Doesn't Load (Data Is Read But Not Used)

### ISSUE THREAD

**Issue by Fronkln** [lewa-j/Unity-Source-Tools#44]
Alright, before posting this issue i 100% checked this out.
In Source, named lights have their own lightmap pages(when a light entity gets turned off, its page is also turned off, that's why)

So i made a test map.

![image](https://user-images.githubusercontent.com/17107072/95890862-00ce4500-0d8d-11eb-8c7f-0caabcdaffed.png)

In the map, there are three lights, light 1 and 2 are named lights so they have their own lightmap page.
Light 3 is a unnamed light.

Loading the map in Source to Unity provides us with only the unnamed light loaded.
![image](https://user-images.githubusercontent.com/17107072/95891006-26f3e500-0d8d-11eb-91a3-7726b45a6e24.png)

Unity Source Tools reports the lightmap pages as 1. 
![image](https://user-images.githubusercontent.com/17107072/95891201-6d494400-0d8d-11eb-8c30-c65ddea89c13.png)

BSP Version: 20 (Source SDK 2013)
[usrc_lightmappages.zip](https://github.com/lewa-j/Unity-Source-Tools/files/5372893/usrc_lightmappages.zip)

**Now here is the most important part. Removing one of the named lights from the BSP reduces lightmap lump which Unity Source tools accurately reports. But it doesn't seem to do anything with the lumps of named lights.**

**Comment by Fronkln** [comment]
Update: Apparently the lightmap data is read but unused

![image](https://user-images.githubusercontent.com/17107072/95898321-8a831000-0d97-11eb-81bc-fa3e3456686a.png)

**Comment by Fronkln** [comment]
On a BSP face, there is a thing called **styles**, this is a byte array of 4. Which is the maximum amount of lights that can affect a face. I'm assuming the styles are not loaded?

**Comment by Fronkln** [comment]
This could be what we need

![image](https://user-images.githubusercontent.com/17107072/95904703-87d8e880-0da0-11eb-8bc9-4aa70149d65e.png)

**Comment by Fronkln** [comment]
![image](https://user-images.githubusercontent.com/17107072/95906125-61b44800-0da2-11eb-8c20-1e677e8a1613.png)
Relevant code?

**Comment by Fronkln** [comment]
255 = empty lightmap style

**Comment by antim0118** [comment]
As I can know, you need to skip lightmap samples for default lightstyle, next is gonna be second lightstyle.
https://developer.valvesoftware.com/wiki/Source_BSP_File_Format#Lighting

**Comment by Fronkln** [comment]
@antim0118 would we be reading them like how we read normal lightmaps?

**Comment by antim0118** [comment]
I think, yes

**Comment by Fronkln** [comment]
We need to find a way to calculate surf size then, if it's already calculated somewhere in the code @lewa-j please let me know.

**Comment by lewa-j** [comment]
Yes. bspface.LightmapTextureSizeInLuxels

**Comment by Fronkln** [comment]
I didn't know that was the surf size the code talks about, thank you.

**Comment by Fronkln** [comment]
![image](https://user-images.githubusercontent.com/17107072/97798857-14a5f200-1c3b-11eb-815b-2cd1d0c8c153.png)
If i am not mistaken, lightofs is divided by 4  in Unity Source tools intentionally to load the default style?

**Comment by Fronkln** [comment]
![image](https://user-images.githubusercontent.com/17107072/97800076-3c995380-1c43-11eb-9f67-9864e83cfc8f.png)
I came up with the following code. but its not completely working. Style 4(default) on each face is completely fine.
I very likely calculated the offset wrong

![image](https://user-images.githubusercontent.com/17107072/97800118-bc272280-1c43-11eb-82c6-fa9a4551829d.png)

![image](https://user-images.githubusercontent.com/17107072/97800126-cea15c00-1c43-11eb-9802-54d48af3fdd7.png)

If anyone wants to take a look at my experimental code lemme know

updated test map: 
[usrc_lightmappages_upd.zip](https://github.com/lewa-j/Unity-Source-Tools/files/5470545/usrc_lightmappages_upd.zip)

**Comment by Fronkln** [comment]
I found out the issue, doing further testing

**Comment by Fronkln** [comment]
Here's the loop i managed to create:

![image](https://user-images.githubusercontent.com/17107072/97805569-f3a7c600-1c67-11eb-9dcd-8d590e7c9810.png)

Output:
![image](https://user-images.githubusercontent.com/17107072/97805582-07ebc300-1c68-11eb-9079-443b4faf13fa.png)

**Comment by Fronkln** [comment]
I wrote a loop to test things further:
![image](https://user-images.githubusercontent.com/17107072/97805648-85173800-1c68-11eb-90cd-83d6674a758b.png)

Output:
![image](https://user-images.githubusercontent.com/17107072/97805655-92ccbd80-1c68-11eb-99c0-99be7b0e0f53.png)


I think we are pretty close to solving this one. We need to combine this with Unity Source Tool's loop to extract all styles from the face. Also ignore the fact that i called them lightmap page in logging it's just the styles

**Comment by lewa-j** [comment]
> If i am not mistaken, lightofs is divided by 4 in Unity Source tools intentionally to load the default style?

It's divided because color is 4 byte long.

**Comment by Fronkln** [comment]
Alright, thank you!

**Comment by Fronkln** [comment]
![image](https://user-images.githubusercontent.com/17107072/97806183-6fefd880-1c6b-11eb-89c9-05490579ff0f.png)
Fixed loop oops we no longer add the same thing

**Comment by Fronkln** [comment]
![image](https://user-images.githubusercontent.com/17107072/97811062-1a2a2900-1c89-11eb-8e4b-53a32465a349.png)
BOOM! It's done! Now we have to write a shader/script that can combine lightmap styles to one

**Comment by Fronkln** [comment]
We need to write a custom shader to additively apply lightmaps to solve this

**Comment by antim0118** [comment]
> BOOM! It's done! Now we have to write a shader/script that can combine lightmap styles to one

You can make two lightmaps colors add/multiply (+ or *) each other from code. I think, that's how it may work in source engine code.

**Comment by Fronkln** [comment]
> You can make two lightmaps colors add/multiply (+ or *) each other from code. I think, that's how it may work in source engine code.

That's true, but calculating the combinations would be a bit of a difficult/expensive task i believe, indeed in Source engine they are applied addittively, that's why i'm talking about writing a shader which can apply textures addittively on top of the base texture
